### ISSUE TITLE
[RFC] better Router profiler panel and debugging

### ISSUE THREAD

**Issue by Tobion** [symfony/symfony#26302]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | no
| BC Break report? | no
| RFC?             | yes
| Symfony version  | 4.1

With the router improvements in SF 4.1, the router profiler panel and `router:match` console command are not reflecting the optimized router matching anymore.
So far those tools output how many routes routes didn't match and why they didn't match before finding the matching route. This suggests that having often-hit routes first is beneficial. This was accurate until SF 4.1. 

But with the recent improvements, the order of routes is pretty much irrelevant to performance. This is because regexes of dynamic routes are combined together. Performance tests that confirm this are in http://developer.happyr.com/symfony-route-test
With PHP 7.2 the order of static routes is also irrelevant because `switch` statements are optimized to a simple hash lookup independent of order.

So the question is how to make the router debugging more useful. What information would you like to get displayed in the router profiler panel? What were you looking for when using the panel? Or did you not use it at all? Please share your experience and ideas in the comments. Thanks

**Comment by Tobion** [comment]
One idea is to show the number of dynamic and static routes.

**Comment by weaverryan** [comment]
I’ve never found anything particularly useful about the router profiler. If it *is* useful, I can only think of it being basically a web version of debug:router as a way to see a list of routes in general.

For router:match I think it’s easier: it should tell you what route is matched, and that’s it. 

In both cases, I’ve never noticed or used any features related to *why* a route matched or didn’t match (except for the router:match feature where it tells you, for example, that there is no match, except for a different http method e.g. POST).

In other words, I think we can just simplify.

**Comment by sebastianblum** [comment]
I had the case that a dynamic route was defined earlier and so my static route doesn‘t work.

Maybe we can add a hint to the profilier that other routes with lower priority exists.

**Comment by sstok** [comment]
@sebastianblum Is this still the case with the new Routing component? If understand correctly static routes are now checked first.

Checking if another route also matches (without giving a URL to mach) _is_ possible but complicated.
In https://github.com/rollerworks/app-sectioning-bundle (planning to change this) I am checking if a pattern also matches another pattern, to produce negative matching regexes for the other prefix (eg. `^/(?!(api|backend)/)`).

Uses https://github.com/icomefromthenet/ReverseRegex
But it's not a 100% fool-proof and can easily break for more complex patterns.

**Comment by sebastianblum** [comment]
@sstok no, that‘s not the case and the new component had the same order/priority imo.

I suggest to display the priority / order of the route in the toolbar and if possible a hint that another route with lower priority also matches. That would be an improvement imo.

**Comment by Xymanek** [comment]
Things I use the routing profiler for (or I want to see there):

- Why route X was not matched and I got 404 (or 405)
- Why route Y was matched instead of (intended) route X
- List of all routes with their requirements (useful when working bundles that add routes)

**Comment by althaus** [comment]
Dunno if this is the right place, but I'd love if the debug console command would be improved. I'm using this all the time and basically never the panel.

In: `console deb:rout user`
Out: The route "user" does not exist.

In: `console deb:cont user`
Out: Select one ...
In: `123`
Out: What I was looking for.

So I always end in doing a grep:

In: `console deb:rout | grep user`
Out: List of matching routes
In: `console deb:rout the_user_route_i_was_looking_for`
Out: What I was looking for.

Should this be a different issue?

Cheers

**Comment by stof** [comment]
Well, the order of routes is still as much important as before regarding the functional part (priority of routes matters in case they are not exclusive).
The fact that the implementation optimizes the way it checks routes does not change that (it is purely internal) but it indeed means that you don't need to be concerned as much as before about optimizing performance with the order of your routes (and so the number of tested routes before match is indeed something we don't care about anymore).

Note however that I never use these features so I don't even know what a useful info would be for the routing panel (maybe it should even be dropped ?)

@althaus you are talking about a different command here. this issue is talking about the `router:match` command, not about `debug:router`
